cli/npm/resolution/common.rs

```diff
@@ -0,0 +1,239 @@
+use deno_core::error::AnyError;
+use deno_graph::semver::Version;
+use deno_graph::semver::VersionReq;
+use once_cell::sync::Lazy;
+
+use super::NpmPackageId;
+use crate::npm::registry::NpmPackageInfo;
+use crate::npm::registry::NpmPackageVersionInfo;
+
+pub static LATEST_VERSION_REQ: Lazy<VersionReq> =
+  Lazy::new(|| VersionReq::parse_from_specifier("latest").unwrap());
+
+pub fn resolve_best_package_version_and_info<'info, 'version>(
+  version_req: &VersionReq,
+  package_info: &'info NpmPackageInfo,
+  existing_versions: impl Iterator<Item = &'version Version>,
+) -> Result<VersionAndInfo<'info>, AnyError> {
+  if let Some(version) = resolve_best_from_existing_versions(
+    version_req,
+    package_info,
+    existing_versions,
+  )? {
+    match package_info.versions.get(&version.to_string()) {
+      Some(version_info) => Ok(VersionAndInfo {
+        version,
+        info: version_info,
+      }),
+      None => {
+        bail!(
+          "could not find version '{}' for '{}'",
+          version,
+          &package_info.name
+        )
+      }
+    }
+  } else {
+    // get the information
+    get_resolved_package_version_and_info(version_req, package_info, None)
+  }
+}
+
+#[derive(Clone)]
+pub struct VersionAndInfo<'a> {
+  pub version: Version,
+  pub info: &'a NpmPackageVersionInfo,
+}
+
+fn get_resolved_package_version_and_info<'a>(
+  version_req: &VersionReq,
+  info: &'a NpmPackageInfo,
+  parent: Option<&NpmPackageId>,
+) -> Result<VersionAndInfo<'a>, AnyError> {
+  if let Some(tag) = version_req.tag() {
+    tag_to_version_info(info, tag, parent)
+  } else {
+    let mut maybe_best_version: Option<VersionAndInfo> = None;
+    for version_info in info.versions.values() {
+      let version = Version::parse_from_npm(&version_info.version)?;
+      if version_req.matches(&version) {
+        let is_best_version = maybe_best_version
+          .as_ref()
+          .map(|best_version| best_version.version.cmp(&version).is_lt())
+          .unwrap_or(true);
+        if is_best_version {
+          maybe_best_version = Some(VersionAndInfo {
+            version,
+            info: version_info,
+          });
+        }
+      }
+    }
+
+    match maybe_best_version {
+      Some(v) => Ok(v),
+      // If the package isn't found, it likely means that the user needs to use
+      // `--reload` to get the latest npm package information. Although it seems
+      // like we could make this smart by fetching the latest information for
+      // this package here, we really need a full restart. There could be very
+      // interesting bugs that occur if this package's version was resolved by
+      // something previous using the old information, then now being smart here
+      // causes a new fetch of the package information, meaning this time the
+      // previous resolution of this package's version resolved to an older
+      // version, but next time to a different version because it has new information.
```

Comment:
Did this change at all? Or are we still gonna have cache busting problems? (https://github.com/denoland/deno/issues/16901)

Reply:
Hasn't changed.